Feature/mypage

JungMinji08 opened the pull request:
- [x] 유저 프로필 UI
- [x] 계정 삭제하는 API 연결
- [x] 내 담글 없을 때 UI
- [x] 내 담글 목록 UI
- [x] 푸쉬 알림 on/off UI
- [x] 프로필 (auth/me) 정보 가져오는 API 연결
- [x] 내 담글 목록 가져오는 API 연결
- [x] 푸시 알림 설정 API 연결
- [ ] 계정 삭제 버튼 눌렀을 때 다이얼로그(삭제/삭제취소)
- [x] 내 담글 목록 UI
- [ ] 담글 Model(presentation layer) 다른 화면에서 공통으로 사용할 수 있도록 수정. MyDamgle(x), Damgle(o)

JungMinji08 committed:
Add `notification: Boolean` field to UserProfile Model

JungMinji08 committed:
Add Get UserProfile Logic

- 사용자 프로필 정보 가져오는 로직 추가
- @GET "v1/auth/me"

JungMinji08 committed:
Add Delete User Logic

- 유저 삭제(탈퇴) 로직 추가

JungMinji08 committed:
Add BackPressInterceptor to HomeScreen

- 홈화면에 뒤로가기 인터셉터 추가
- Splash 화면에서 다른 화면으로 이동할 때 네비게이션 stack clear

JungMinji08 committed:
Refactor MyPageViewModel Structure & Connect MyDamgleList API

- MyPageViewModel 구조 리팩토링
- MyDamgleList API 연결

JungMinji08 committed:
Fix Hard Coding String

JungMinji08 committed:
Add MyPage Empty Screen when MyDamgleList is empty

- 마이페이지 MyDamgleList가 비었을 때 화면 추가

JungMinji08 committed:
Add Patch Notification Allow State

- 알림 허용 상태 패치하는 API 연결

JungMinji08 committed:
Leave TODO comments

JungMinji08 committed:
Add Compose Preview (When MyDamgle List is empty)

JungMinji08 committed:
Change Orange Gradient Background Image Resource

JungMinji08 committed:
Add MyPage TabBar Animation

JungMinji08 commented:
계정 삭제 눌렀을 때 다이얼로그로 한번 더 확인하는거랑 Damgle Model 공통으로 빼는거는 나중에 다른 PR들 합치고 수정하겠습니다...!

dahyun1226 commented on app/src/main/java/com/mashup/damgledamgle/presentation/feature/mypage/MyPageViewModel.kt:
```diff
@@ -18,18 +28,81 @@ import javax.inject.Inject
+    private val userProfileMapper: UserProfileMapper,
+    private val damgleMapper: DamgleMapper,
+): ViewModel() {
+    val myDamgleListState = MutableStateFlow<ViewState<List<DamgleModel>>>(ViewState.Loading)
+    val userProfileState = MutableStateFlow<ViewState<UserProfileModel>>(ViewState.Loading)
```
이 둘이 같은 ViewState로 가져가진다면 (combine으로 묶어주는게 그 이유니까)
각각이 따로 ViewState를 가져가야하나 의문이 드네요 
한개로 묶어줄 수 있을거 같은?

dahyun1226 commented on app/src/main/java/com/mashup/damgledamgle/presentation/feature/mypage/MyPageViewModel.kt:
```diff
@@ -18,18 +28,81 @@ import javax.inject.Inject
+    private fun getMyDamgleList() {
+        viewModelScope.launch {
+            when (val result = getMyDamgleListUseCase()) {
+                is NetworkResponse.Success -> myDamgleListState.emit(
+                    ViewState.Success(
+                        damgleMapper.mapToModel(result.data)
+                    )
+                )
+                is NetworkResponse.Error -> myDamgleListState.emit(ViewState.Error(result.exception.message.toString()))
+            }
+        }
```
매번 이렇게 Error Success 핸들링 하는것보다는 이렇게 하는게 나을지도?
(엄청 러프하게 짰고 훨씬 더 깔끔하고 나은 방식, 예제가 있을거 같음!)
~~~kotlin
// 대충 NetworkResponse에 넣기 (실제 쓸때는 실패도 넣어야겠죠?)
fun <T> CoroutineScope.launchWithNetworkResult(networkResponse: NetworkResponse<T>, successBlock: suspend CoroutineScope.(T) -> Unit) {
    if (networkResponse is NetworkResponse.Success<*>) {
        launch {
            successBlock(networkResponse.data as T)
        }
    }
}
// 뷰모델에서 사용
viewModelScope.launch {
    launchWithNetworkResult(
        getUserProfileUserCase()
    ) {
        userProfileState.emit(
            ViewState.Success(userProfileMapper.mapToModel(it))
        )
    }
 }
~~~

dahyun1226 replied:
suspendOnSuccess, suspendOnError 등으로 짜도 괜찮을거 같다는 생각!
이거 잘 짜보는것도 재미있을거 같아서 시간나면 해봐야겠다고 생각이 드네요

dahyun1226 replied:
NetworkResponse가 프레젠트 단으로 와야하는가? 도 약간의문이 들긴 합니다

JungMinji08 replied:
오호 이제 살짝 이해한듯...? 👍👍👏👏
NetworkResponse를 프레젠트까지 안보내려면 이런식으로 방법이 있었구나... 
MyPageViewModel코드에 적용해보긴했는데 내가 제대로 이해하고 한건지 확인부탁해염
일단 MyPage쪽에 넣은거 괜찮으면 OnBoardingViewModel쪽에도 적용할게여 🥸

dahyun1226 replied:
요거는 NetworkResponse 프레젠트단까지 보내는거랑은 약간 다르긴해요
NetworkResponse를 프레젠트단에서 사용할때는 이렇게 하면 더 좋지 않을까? 정도였고
프레젠트 단에서 사용하지 않으려면 repository단에서 아예 바꿔서 전달 해줄 수 있지 않을까 싶은??

JungMinji08 replied:
@dahyun1226 
e87d1e4ab880b60837f2e32cab1416600acb626e 커밋으로 수정했슴다..~!

dahyun1226 commented on app/src/main/java/com/mashup/damgledamgle/presentation/feature/mypage/PushSettingItem.kt:
```diff
@@ -26,8 +29,11 @@ import com.mashup.damgledamgle.ui.theme.*
 @Composable
-fun PushSettingItem() {
-    val pushAllowState = remember { mutableStateOf(false) }
+fun PushSettingItem(
+    notification: Boolean,
+) {
+    val viewModel: MyPageViewModel = hiltViewModel()
+    val pushAllowState = remember { mutableStateOf(notification) }
```
뷰모델 가질 필요보다는 람다로 이벤트 넘겨주는게 좋아보입니다

JungMinji08 committed:
Refactor PushSettingItem

- PushSettingItem에서 뷰모델을 가지지 않고 부모로부터 람다식을 전달받아서 사용

JungMinji08 committed:
Add <T>CoroutineScope.launchWithNetworkResult() function

- NetworkResponse 값에 따라 이벤트를 핸들링하는 확장함수 생성
- <T>CoroutineScope.launchWithNetworkResult()
- @param networkResponse: NetworkResponse<T>
- @param suspendOnSuccess: suspend CoroutineScope.(T) -> Unit
- @param suspendOnError: suspend CoroutineScope.(Exception) -> Unit,

JungMinji08 committed:
Rename NetworkResponse to Result

- 데이터 리퀘스트의 결과를 Success/Error로 표시하는 NetworkResponse 클래스의 이름을 Result로 변경
- 각각 Result.Success, Result.Error 로 사용
